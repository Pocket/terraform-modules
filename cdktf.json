{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
