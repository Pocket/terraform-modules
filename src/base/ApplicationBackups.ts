import { Construct } from 'constructs';
import { TerraformMetaArguments } from 'cdktf';
import {
  BackupPlanRule,
  BackupPlan,
} from '@cdktf/provider-aws/lib/backup-plan';
import {
  BackupSelection,
  BackupSelectionSelectionTag,
} from '@cdktf/provider-aws/lib/backup-selection';
import { BackupVault } from '@cdktf/provider-aws/lib/backup-vault';
import { BackupVaultPolicy } from '@cdktf/provider-aws/lib/backup-vault-policy';

export interface ApplicationBackupProps extends TerraformMetaArguments {
  name: string;
  kmsKeyArn: string;
  prefix: string;
  accountId: string;
  vaultPolicy: string;
  backupPlans: {
    name: string;
    resources: string[];
    rules: Omit<BackupPlanRule, 'targetVaultName'>[];
    selectionTag: BackupSelectionSelectionTag[];
  }[];
  tags?: { [key: string]: string };
}

export class ApplicationBackup extends Construct {
  public backupPlan: BackupPlan;
  public backupSelection: BackupSelection;
  public backupPlanRule: BackupPlanRule;
  private static vault: BackupVault;

  constructor(
    scope: Construct,
    name: string,
    private config: ApplicationBackupProps,
  ) {
    super(scope, name);

    const vault = new BackupVault(this, 'backup-vault', {
      name: `${config.prefix}-${config.name}`,
      kmsKeyArn: config.kmsKeyArn,
      tags: config.tags,
      provider: config.provider,
    });

    new BackupVaultPolicy(this, 'backup-vault-policy', {
      backupVaultName: vault.name,
      policy: config.vaultPolicy,
      provider: config.provider,
    });

    config.backupPlans.forEach((plan) => {
      const backupPlan = new BackupPlan(this, 'backup-plan', {
        name: plan.name,
        rule: plan.rules.map((rule) => ({
          ...rule,
          targetVaultName: vault.name,
        })),
        tags: config.tags,
        provider: config.provider,
      });

      new BackupSelection(this, 'backup-selection', {
        name: `${config.prefix}-Backup-Selection`,
        planId: backupPlan.id,
        iamRoleArn: `arn:aws:iam::${config.accountId}:role/service-role/AWSBackupDefaultServiceRole`,
        resources: plan.resources,
        selectionTag: plan.selectionTag,
        provider: config.provider,
      });
    });
  }
}
