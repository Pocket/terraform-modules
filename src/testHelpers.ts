import { Resource, TerraformStack } from 'cdktf';
import { Construct } from 'constructs';

// used to test individual static functions within component classes
// generates a basic/empty construct Resource to provide context
export class TestResource extends Resource {
  constructor(scope: TerraformStack | Construct, name: string) {
    super(scope, name);
  }
}
