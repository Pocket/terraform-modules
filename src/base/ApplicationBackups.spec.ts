import { Testing } from 'cdktf';
import { TestResource } from '../testHelpers';
import { ApplicationBackup } from './ApplicationBackups';

describe('ApplicationBackup', () => {
    it('renders vault with plans without tags', () => {
        const synthed = Testing.synthScope((stack ) => {
            new ApplicationBackup(stack, 'testBackup', {
                name: 'test-',
                kmsKeyArn: 'arn:aws:kms:us-east-1:1234567890:key/mrk-1234',
                prefix: 'TEST',
                accountId: '1234567890',
                vaultPolicy: '{"Version": "2012-10-17","Statement": [{"Effect": "Allow","Action": "backup:CopyIntoBackupVault","Resource": "*","Principal": "*","Condition": {"StringEquals": {"aws:PrincipalOrgID": ["o-1234567890"]}}}]}',
                backupPlans: [
                    {
                        name: 'TestPlan',
                        resources: ['arn:aws:rds:us-east-1:123456790:db:test'],
                        rules: [{
                            ruleName: 'TestBackupRule',
                            schedule: 'cron( 0 5 ? * * *)'
                        }],
                        selectionTag: [{
                            key: 'backups',
                            type: 'STRINGEQUALS',
                            value: 'True'
                          }]
                    }
                ]
            })
        })
    })

    it('renders vault with plans with tags', () => {
        const synthed = Testing.synthScope((stack ) => {
            new ApplicationBackup(stack, 'testBackup', {
                name: 'test-',
                kmsKeyArn: 'arn:aws:kms:us-east-1:1234567890:key/mrk-1234',
                prefix: 'TEST',
                accountId: '1234567890',
                vaultPolicy: '{"Version": "2012-10-17","Statement": [{"Effect": "Allow","Action": "backup:CopyIntoBackupVault","Resource": "*","Principal": "*","Condition": {"StringEquals": {"aws:PrincipalOrgID": ["o-1234567890"]}}}]}',
                backupPlans: [
                    {
                        name: 'TestPlan',
                        resources: ['arn:aws:rds:us-east-1:123456790:db:test'],
                        rules: [{
                            ruleName: 'TestBackupRule',
                            schedule: 'cron( 0 5 ? * * *)'
                        }],
                        selectionTag: [{
                            key: 'backups',
                            type: 'STRINGEQUALS',
                            value: 'True'
                          }]
                    }
                ],
                tags: { name: 'thedude',
                        hobby: 'bowling'
                }
            })
        })
    })

});