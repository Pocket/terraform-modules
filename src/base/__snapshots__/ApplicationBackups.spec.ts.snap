// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationBackup renders vault with plans with tags 1`] = `
"{
  "resource": {
    "aws_backup_plan": {
      "testBackup_backup-plan_AB394A83": {
        "name": "TestPlan",
        "rule": [
          {
            "rule_name": "TestBackupRule",
            "schedule": "cron( 0 5 ? * * *)",
            "target_vault_name": "\${aws_backup_vault.testBackup_backup-vault_C40F1BAC.name}"
          }
        ],
        "tags": {
          "hobby": "bowling",
          "name": "thedude"
        }
      }
    },
    "aws_backup_selection": {
      "testBackup_backup-selection_67969F7D": {
        "iam_role_arn": "arn:aws:iam::1234567890:role/service-role/AWSBackupDefaultServiceRole",
        "name": "prefix-Backup-Selection",
        "plan_id": "\${aws_backup_plan.testBackup_backup-plan_AB394A83.id}",
        "resources": [
          "arn:aws:rds:us-east-1:123456790:db:test"
        ],
        "selection_tag": [
          {
            "key": "backups",
            "type": "STRINGEQUALS",
            "value": "True"
          }
        ]
      }
    },
    "aws_backup_vault": {
      "testBackup_backup-vault_C40F1BAC": {
        "kms_key_arn": "arn:aws:kms:us-east-1:1234567890:key/mrk-1234",
        "name": "prefix-name",
        "tags": {
          "hobby": "bowling",
          "name": "thedude"
        }
      }
    },
    "aws_backup_vault_policy": {
      "testBackup_backup-vault-policy_A278B298": {
        "backup_vault_name": "\${aws_backup_vault.testBackup_backup-vault_C40F1BAC.name}",
        "policy": "{\\"Version\\": \\"2012-10-17\\",\\"Statement\\": [{\\"Effect\\": \\"Allow\\",\\"Action\\": \\"backup:CopyIntoBackupVault\\",\\"Resource\\": \\"*\\",\\"Principal\\": \\"*\\",\\"Condition\\": {\\"StringEquals\\": {\\"aws:PrincipalOrgID\\": [\\"o-1234567890\\"]}}}]}"
      }
    }
  }
}"
`;

exports[`ApplicationBackup renders vault with plans without tags 1`] = `
"{
  "resource": {
    "aws_backup_plan": {
      "testBackup_backup-plan_AB394A83": {
        "name": "TestPlan",
        "rule": [
          {
            "rule_name": "TestBackupRule",
            "schedule": "cron( 0 5 ? * * *)",
            "target_vault_name": "\${aws_backup_vault.testBackup_backup-vault_C40F1BAC.name}"
          }
        ]
      }
    },
    "aws_backup_selection": {
      "testBackup_backup-selection_67969F7D": {
        "iam_role_arn": "arn:aws:iam::1234567890:role/service-role/AWSBackupDefaultServiceRole",
        "name": "prefix-Backup-Selection",
        "plan_id": "\${aws_backup_plan.testBackup_backup-plan_AB394A83.id}",
        "resources": [
          "arn:aws:rds:us-east-1:123456790:db:test"
        ],
        "selection_tag": [
          {
            "key": "backups",
            "type": "STRINGEQUALS",
            "value": "True"
          }
        ]
      }
    },
    "aws_backup_vault": {
      "testBackup_backup-vault_C40F1BAC": {
        "kms_key_arn": "arn:aws:kms:us-east-1:1234567890:key/mrk-1234",
        "name": "prefix-name"
      }
    },
    "aws_backup_vault_policy": {
      "testBackup_backup-vault-policy_A278B298": {
        "backup_vault_name": "\${aws_backup_vault.testBackup_backup-vault_C40F1BAC.name}",
        "policy": "{\\"Version\\": \\"2012-10-17\\",\\"Statement\\": [{\\"Effect\\": \\"Allow\\",\\"Action\\": \\"backup:CopyIntoBackupVault\\",\\"Resource\\": \\"*\\",\\"Principal\\": \\"*\\",\\"Condition\\": {\\"StringEquals\\": {\\"aws:PrincipalOrgID\\": [\\"o-1234567890\\"]}}}]}"
      }
    }
  }
}"
`;
